File: Python/number-of-valid-subarrays.py
# Time:  O(n)
# Space: O(n)

class Solution(object):
    def validSubarrays(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        result = 0
        s = []
        for num in nums:
            while s and s[-1] > num:
                s.pop()
            s.append(num);
            result += len(s)
        return result

File: C++/number-of-valid-subarrays.cpp
// Time:  O(n)
// Space: O(n)

class Solution {
public:
    int validSubarrays(vector<int>& nums) {
        int result = 0;
        vector<int> s;
        for (const auto& num : nums) {
            while (!s.empty() && s.back() > num) {
                s.pop_back();
            }
            s.emplace_back(num);
            result += s.size();
        }
        return result;
    }
};
